feat!: add adaptor output path mapping and overrides (#55)

gp-dan opened the pull request:
### What was the problem/requirement? (What/Why)

#55 adaptor not path mapping or overriding output path

### What was the solution? (How)

- Add get_output_paths() to Scene to return expanded default and multi-pass paths
- Add output and multi-pass paths to parameters, init data, ui and adaptor actions
- Remove DEADLINE_PYTHON_PATH to ensure proper loading of win32 modules

### What is the impact of this change?

User can override output path and asset outputs are detected and downloadable

### How was this change tested?

- Submitted scene with default settings Override Output Path disabled, the path was mapped by adaptor and files downloaded
- Submitted scene with Override Output Path enabled and augmented to a different name, the path was mapped by adaptor and files downloaded
 
### Was this change documented?

No

### Is this a breaking change?

Potentially -- logic relying on DEADLINE_PYTHON_PATH is removed so this may impact installation instructions 

Manually adding modules from an arbitrary DEADLINE_PYTHON_PATH location failed to properly boostrap more complex modules like win32 with extra DLLs (even when calling os.add_dll_directory)

Installing the modules to a location expected by the embedded python (such as `C:\Users\<user>\AppData\Roaming\Maxon\python\python311\libs`) solves this issue and makes the plugin a more idiomatic Cinema4D extension.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

gp-dan committed:
fix: adaptor output path mapping and overrides (#55)

Signed-off-by: dan <dan@gunpowder.tech>

karthikbekalp commented on deadline_cloud_extension/DeadlineCloud.pyp:
```diff
@@ -4,12 +4,7 @@ import sys
 sys.path.insert(0, os.path.join(root, 'modules'))
 
-if 'cinema4d_submitter' not in sys.modules.keys():
-    python_path = os.getenv('DEADLINE_CLOUD_PYTHONPATH')
```
Considering this has an impact on the environment variable that we set, can we update the instructions as well? I mean [over here](https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/blob/f26c9ac7f213f6be9719226bfcc8965caf328d6c/DEVELOPMENT.md?plain=1#L41)

karthikbekalp replied:
Thanks for updating!

karthikbekalp commented on src/deadline/cinema4d_adaptor/Cinema4DAdaptor/adaptor.py:
```diff
@@ -26,7 +26,7 @@ class Cinema4DNotRunningError(Exception):
 
 
-_FIRST_CINEMA4D_ACTIONS = ["scene_file", "take"]
+_FIRST_CINEMA4D_ACTIONS = ["scene_file", "take", "output_path", "multi_pass_path"]
```
Nit: Should it be `multi_pass_output_path` instead so that it signifies that its an output path?

gp-dan replied:
My experience is the term "multi-pass" is unambiguously an output path in Cinema4D. The c4d api uses the handle `RDATA_MULTIPASS_FILENAME`. We can certainly change this line but it would also mean changing the respective action method, the adaptor schema, the init data and the template --  and probably the render settings data class and the ui widget for consistency sake too.

karthikbekalp replied:
Ok thanks for the clarifications!

karthikbekalp commented on src/deadline/cinema4d_submitter/cinema4d_render_submitter.py:
```diff
@@ -244,10 +246,12 @@ def _show_submitter(parent=None, f=Qt.WindowFlags()):
 
     # Load the sticky settings
-    render_settings.load_sticky_settings(Scene.name())
+    # render_settings.load_sticky_settings(Scene.name())
```
Hmm, any reason we commented this line out? Is it not required anymore?

gp-dan replied:
My mistake, thanks for catching this

karthikbekalp replied:
No problem.

karthikbekalp commented on src/deadline/cinema4d_submitter/scene.py:
```diff
@@ -170,6 +174,43 @@ def get_render_data(doc=None, take=None):
+            take_data = doc.GetTakeData()
+            take = take_data.GetCurrentTake()
+        render_data = Scene.get_render_data(doc=doc, take=take)
+        rbc = render_data.GetDataInstance()
```
Nit: Can we have better variables names? It is not intuitive as what rbc or rpd mean. :)

gp-dan replied:
I agree names could be better but they match the dict key names in the c4d api here which also uses [_rBc](https://developers.maxon.net/docs/py/2024_0_0a/modules/c4d.modules/tokensystem/index.html#:~:text=_rData%3A%20RenderData-,_rBc,-%3A%20BaseContainer) - I could rename it `render_base_container_instance` ?

karthikbekalp replied:
That would make more sense, thanks!

karthikbekalp reviewed (changes requested):
Thanks for your contribution! On the whole this PR looks really good, but I've added few comments that require some clarifications/improvements. 

Also, we need to add unit tests for this.

gp-dan committed:
fix: sticky settings, install docs and lint issue

- fix sticky settings regression
- update install instructions for python module path
- improved variable name in scene module
- run hatch fmt to fix lint issue

Signed-off-by: dan <dan@gunpowder.tech>

karthikbekalp commented on DEVELOPMENT.md:
````diff
@@ -38,15 +38,14 @@ hatch run all:test
 ```
 # deadline-cloud lib with pyside
-export DEADLINE_CLOUD_PYTHONPATH="/path/to/deadline-cloud/site-packages"
+export C4DPYTHONPATH311 "/path/to/deadline-cloud/site-packages"
````
Missing `=`? Also this may have some merge conflicts with[ this PR](https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/pull/59/files).  Same issue below

karthikbekalp replied:
Also any reason this is suffixed with 311? I was able to get it working on a machine with Python 3.12 as well.

gp-dan replied:
Yes, thanks, missing = is a typo.
Reason for suffix is dev docs seem to indicate C4DPYTHONPATH is for python 2.7 only, I wouldn't be surprised if 3.12 worked too, especially pure python modules: https://developers.maxon.net/docs/py/2024_0_0a/manuals/manual_py_libraries.html#custom-interpreter-bound-search-paths:~:text=%3C%3D%20C4D_VERSION)-,C4DPYTHONPATH311,-Note%20To%20register

karthikbekalp replied:
Thanks for your clarification!

karthikbekalp commented on DEVELOPMENT.md:
````diff
@@ -68,16 +66,6 @@ client.
 export DEADLINE_CINEMA4D_EXE="/opt/maxon/cinema4dr2024.200/bin/c4d"
-```
-
-### Windows
````
Hmm, so we don't need to manually install pywin32 anymore? But wouldn't this raise module not found errors by the adaptor?

gp-dan replied:
when I pip install deadline-cloud-for-cinema-4d to a interpreter search path supported by Cinema 4D it installs the pywin32 dependency too, both submitter and adaptor load it

karthikbekalp replied:
Hmm, I've had issues with this error when I don't install pywin32 separately. But I'll investigate this on my side. Thanks for the tip. 

```
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)Traceback (most recent call last):
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)  File "C:\ProgramData\Amazon\OpenJD\session-1e008b980032473da35ef8f59bb9942e09l386vr\.env\Lib\site-packages\deadline\cinema4d_adaptor\Cinema4DClient\plugin\DeadlineCloudClient.pyp", line 25, in <module>
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)    from deadline.cinema4d_adaptor.Cinema4DClient.cinema4d_client import main
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)  File "C:\ProgramData\Amazon\OpenJD\session-1e008b980032473da35ef8f59bb9942e09l386vr\.env\Lib\site-packages\deadline\cinema4d_adaptor\Cinema4DClient\__init__.py", line 3, in <module>
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)    from .cinema4d_client import Cinema4DClient
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)  File "C:\ProgramData\Amazon\OpenJD\session-1e008b980032473da35ef8f59bb9942e09l386vr\.env\Lib\site-packages\deadline\cinema4d_adaptor\Cinema4DClient\cinema4d_client.py", line 9, in <module>
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)    from openjd.adaptor_runtime_client import (
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)  File "C:\ProgramData\Amazon\OpenJD\session-1e008b980032473da35ef8f59bb9942e09l386vr\.env\Lib\site-packages\openjd\adaptor_runtime_client\__init__.py", line 19, in <module>
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)    from .win_client_interface import WinClientInterface as ClientInterface  # type: ignore
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)  File "C:\ProgramData\Amazon\OpenJD\session-1e008b980032473da35ef8f59bb9942e09l386vr\.env\Lib\site-packages\openjd\adaptor_runtime_client\win_client_interface.py", line 12, in <module>
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)    from .named_pipe.named_pipe_helper import NamedPipeHelper
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)  File "C:\ProgramData\Amazon\OpenJD\session-1e008b980032473da35ef8f59bb9942e09l386vr\.env\Lib\site-packages\openjd\adaptor_runtime_client\named_pipe\named_pipe_helper.py", line 7, in <module>
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)    import win32file
2024/10/25 07:48:35+00:00 ADAPTOR_OUTPUT: STDOUT: (null)ModuleNotFoundError: No module named 'win32file'
```

karthikbekalp reviewed (changes requested):
Thanks for your contribution! Added a few comments for clarification.

karthikbekalp commented on deadline_cloud_extension/DeadlineCloud.pyp:
```diff
@@ -4,12 +4,7 @@ import sys
-if 'cinema4d_submitter' not in sys.modules.keys():
-    python_path = os.getenv('DEADLINE_CLOUD_PYTHONPATH')
-    python_paths = python_path.split(os.pathsep)
-    for n in python_paths:
```
I was testing this and I got errors like these 

```
Traceback (most recent call last):
  File "D:\C4D_Submitter_Externsion\DeadlineCloud.pyp", line 8, in <module>
import deadline.cinema4d_submitter
ModuleNotFoundError:No module named 'deadline'
```

gp-dan replied:
Have you installed the modules to a supported interpreter path? https://developers.maxon.net/docs/py/2024_0_0a/manuals/manual_py_libraries.html#python-interpreter-bound-search-paths

karthikbekalp replied:
Hmm, using the `C4DPYTHONPATH311` fixed my issues.

gp-dan committed:
fix: take token expansion

Signed-off-by: dan <dan@gunpowder.tech>

gp-dan committed:
fix: typo in docs for example environment script

Signed-off-by: dan <dan@gunpowder.tech>

karthikbekalp reviewed (dismissed):
Thanks for your contributions! The changes look good to me.

epmog commented on src/deadline/cinema4d_submitter/scene.py:
```diff
@@ -170,6 +174,40 @@ def get_render_data(doc=None, take=None):
         return render_data
 
+    @staticmethod
+    def get_output_paths(take=None) -> str:
```
Are we duplicating all of `get_output_directories` here? Looks like the main difference is that we want the specific values rather than having the two of them in a set? Seems like we could completely remove the old usage, and just use the results from this function?

gp-dan replied:
hi @epmog , get_output_directories() is for fully expanded directories and called to build the asset references structure --  in contrast get_output_paths() is only for getting the save image and save multi-pass "path" values (not directory values) which can be subtle and often don't have a fully expanded name, for example things line frame and extension are finalized by c4d during render time. They could be combined to reduce lines of code but you might end up with more complexity in one function that must navigate corner cases of 2 different use cases. It's a subjective call on net cost of code maintenance - my instinct is to wait until later to factor out any common code as combining them now might have unintended side effects that need more testing.

epmog replied:
Hmm... If I do a diff between the methods, that difference in functionality doesn't exist? It looks like one is missing pulling the default take data, we may want to add that? `get_output_paths` appears to otherwise return the exact same data as `get_output_directories`, just formatted differently (ie. dirname for asset references and in a set, but the underlying value being access is the same). Which if we're adding them as PATH + OUTPUT job parameters to the template will automatically handle that. 

Let me know if I'm missing something there.

If there is an actual difference between the two, then I ask that we document what that is as a doc-string, otherwise I'd err on having the one function and if we're still manually adding the asset reference for the outputs, then we can take the dirname of the result of one function. 

```diff
-    def get_output_directories(render_data=None, take=None) -> set[str]:
-        """
-        Returns a list of directories files will be output to.
-        """
+    def get_output_paths(take=None) -> str:
         doc = c4d.documents.GetActiveDocument()
         doc_path = doc.GetDocumentPath()
-        if not take:
-            take_data = doc.GetTakeData()
-            take = take_data.GetCurrentTake()
+
+
+        
         render_data = Scene.get_render_data(doc=doc, take=take)
-        rbc = render_data.GetDataInstance()
+        render_base_container_instance = render_data.GetDataInstance()
         rpd = {
             "_doc": doc,
             "_rData": render_data,
-            "_rBc": rbc,
+            "_rBc": render_base_container_instance,
             "_frame": doc.GetTime().GetFrame(doc.GetFps()),
         }
         if take:
             rpd["_take"] = take
-        image_paths = set()
+        default_out = ""
+        multi_out = ""
         if render_data[c4d.RDATA_SAVEIMAGE]:
             path = render_data[c4d.RDATA_PATH]
             xpath = c4d.modules.tokensystem.FilenameConvertTokens(path, rpd)
@@ -25,7 +23,7 @@
                 if xpath.startswith("./"):
                     xpath = xpath[2:]
                 xpath = os.path.join(doc_path, xpath)
-            image_paths.add(os.path.dirname(os.path.normpath(xpath)))
+            default_out = os.path.normpath(xpath)
         if render_data[c4d.RDATA_MULTIPASS_SAVEIMAGE]:
             path = render_data[c4d.RDATA_MULTIPASS_FILENAME]
             xpath = c4d.modules.tokensystem.FilenameConvertTokens(path, rpd)
@@ -33,5 +31,5 @@
                 if xpath.startswith("./"):
                     xpath = xpath[2:]
                 xpath = os.path.join(doc_path, xpath)
-            image_paths.add(os.path.dirname(os.path.normpath(xpath)))
-        return image_paths
\ No newline at end of file
+            multi_out = os.path.normpath(xpath)
+        return default_out, multi_out
\ No newline at end of file

```

Interestingly... we're previously completely overriding the render_data without using the parameter in get_output_directories... but that's just a musing.

gp-dan replied:
howdy @epmog it's common in other submitters to have 2 methods for this - for example in maya [get_output_directories()](https://github.com/aws-deadline/deadline-cloud-for-maya/blob/mainline/src/deadline/maya_submitter/scene.py#L119) and [output_path()](https://github.com/aws-deadline/deadline-cloud-for-maya/blob/mainline/src/deadline/maya_submitter/scene.py#L136) are the equivalent of our get_output_directories() and get_output_paths() respectively -- the only difference is we return a tuple default_out, mulit_out here. It's a feature that get_output_paths() doesn't fall back to main take, if it did the output mapping of each take would be overridden with the main take path - this is what I mean about 2 use cases needing 2 different functions, combining them can open the door to side effects that break other features.  I'll add a doc string to get_output_paths() to make this clearer.

epmog reviewed (changes requested):
Aside from the one comment to discuss and perhaps de-dupe, this will have to be a breaking change (signified by `feat!`) due to changing how the installation works and changing the adaptor interface (a new submitter requires a new adaptor in this case)

Otherwise, LGTM

gp-dan committed:
fix!: return type signature and add docstring on get_output_paths()

Remove unused output_path()

Signed-off-by: dan <dan@gunpowder.tech>

karthikbekalp reviewed (approved):
Thanks for your contribution. Approving the changes.

karthikbekalp committed:
Merge branch 'mainline' into map-output-gh55
